### PR TITLE
[Chore] Add benchmark for the UMAP negative-sampling branch

### PR DESCRIPTION
**Pull request by huguesva**
## Verdict

**REFUTE** — packing the UMAP negative pairs down to the per-row active budget
is a real but scale-dependent saving that clears neither campaign threshold
across the target range, and it costs reproducibility. No production change is
proposed. This PR adds the benchmark that produced the evidence so the decision
stays re-checkable.

## Hypothesis

`UMAP._compute_repulsive_gradients` draws `n_negatives = negative_sample_rate *
n_neighbors` negatives for every row of the local chunk, evaluates the full
`(chunk, n_negatives)` rectangle, and then zeroes every column past that row's
per-iteration budget `min(active_positive_edges * negative_sample_rate,
n_negatives)`. Computing only the surviving pairs should cut repulsive kernel
time and peak memory, especially later in optimization when the edge schedule
has thinned out.

The campaign also asked to test a cheaper alternative first: truncating the
rectangle to the widest per-row budget of the step, instead of a packed kernel.

## Benchmark design

- Baseline commit: `697d7cb` (`upstream/main`), also the base of this branch.
- Candidate commits: none in `torchdr/`. The two candidate kernels live in the
  added benchmark only.
- Hardware and world size: 1 × NVIDIA B200 (183 GB), driver 580.126.20,
  torch 2.8.0+cu128, `backend="faiss"`, float32, world size 1.
- Datasets: Macosko 2015 (44,808 × 50) and 10x mouse Zheng (1,306,127 × 50 and a
  200,000-row subsample), both the PCA-50 matrices used by
  `benchmarks/benchmark_umap_single_cell.py`.
- `n_neighbors=30` (so `n_negatives=150`), `max_iter=500`, `random_state=0`.
- Repetitions: 3 measured `fit_transform` runs per arm after a 20-iteration
  warm-up; median reported with `[min, max]`. Kernel times come from
  preallocated CUDA events with no host synchronization inside the loop; wall
  times are bracketed by `torch.cuda.synchronize()`.
- Three arms in the same process, on the same device-resident data, same seed,
  same iteration count, so the only difference is the repulsive kernel:
  `baseline` (shipped), `fused` (control: rectangular, but the difference tensor
  is materialized once instead of twice — bitwise identical), `packed`
  (candidate).
- Exact command (per configuration; SLURM jobs `20736835`, `20736277`):

  ```
  python benchmarks/umap_negative_sampling.py \
      --dataset zheng --n-samples 200000 --n-neighbors 30 \
      --max-iter 500 --reps 3 --quality-k 15 \
      --arms baseline,fused,packed --check
  ```

## Results

`step` and `repulsive` are per-iteration CUDA-event medians. `peak step` is the
exact peak allocated bytes of a training step (max of the two phase peaks, each
measured from its own `reset_peak_memory_stats`).

**Macosko, n = 44,808, k = 30** (job `20736835`)

| arm | wall s | Δ wall | step ms | repulsive ms | peak step MB | peak repulsive MB | preservation@15 |
|---|---:|---:|---:|---:|---:|---:|---:|
| baseline | 1.553 [1.547, 1.555] | — | 2.599 | 0.654 | 815 | 574 | 0.147044 |
| fused | 1.484 [1.484, 1.485] | −4.4% | 2.465 | 0.519 | 814 | 600 | 0.147044 |
| packed | 1.518 [1.517, 1.522] | −2.3% | 2.625 | 0.526 | 761 | 466 | 0.148612 |

**Zheng subsample, n = 200,000, k = 30** (job `20736277`)

| arm | wall s | Δ wall | step ms | repulsive ms | peak step MB | peak repulsive MB | preservation@15 |
|---|---:|---:|---:|---:|---:|---:|---:|
| baseline | 7.662 [7.659, 7.665] | — | 13.800 | 2.885 | 5279 | 3524 | 0.078048 |
| fused | 7.398 [7.386, 7.403] | −3.4% | 13.267 | 2.351 | 5279 | 3644 | 0.078048 |
| packed | 6.891 [6.890, 6.898] | **−10.1%** | 12.816 | 1.821 | 5035 | 3041 | 0.078351 |

**Zheng full, n = 1,306,127, k = 30** (job `20736277`)

| arm | wall s | Δ wall | step ms | repulsive ms | peak step MB | peak repulsive MB | preservation@15 |
|---|---:|---:|---:|---:|---:|---:|---:|
| baseline | 77.774 [77.735, 82.805] | — | 125.369 | 19.728 | 51722 | 33250 | 0.029727 |
| fused | 78.173 [76.109, 79.014] | +0.5% | 121.982 | 16.329 | 51753 | 34065 | 0.029727 |
| packed | 72.488 [72.447, 72.512] | −6.8% | 118.217 | 12.467 | 50165 | 30125 | 0.029310 |

An independent earlier job (`20734878`, same measurement code) reproduces the
candidate deltas: macosko −1.3%, Zheng-200k −10.5%, Zheng-1.3M −6.9% wall;
repulsive-branch −20.8% / −37.0% / −36.8%.

**Density, and the cheaper alternative**

| dataset | packed pairs / iter | rectangular pairs / iter | ratio | widest per-row budget (median) | iterations saturating `n_negatives` |
|---|---:|---:|---:|---:|---:|
| macosko 44,808 | 1,998,860 | 6,721,200 | 29.7% | 400 | 99.80% |
| zheng 200,000 | 8,846,277 | 30,000,000 | 29.5% | 525 | 99.80% |
| zheng 1,306,127 | 58,170,182 | 195,919,050 | 29.7% | 825 | 100.00% |

The density is stable at ~30% from the first hundred iterations onward — there
is no "late training gets sparser" effect to exploit. And the widest per-row
budget is 400–825 against `n_negatives = 150`, so **truncating the rectangle to
the step maximum saves nothing in 99.8–100% of iterations**. Hub rows in the
symmetrized k-NN graph accumulate far more than `n_neighbors` active positive
edges and always saturate the cap. The cheap alternative is a dead end; only a
genuinely packed kernel can capture the 70%.

## Correctness and quality

The `--check` probe replays every arm against one identical model state
(embedding, `mask_affinity_in_`, `neg_indices_`, exclusion tables) at iterations
0, 1, 250 and 499. Over the 16 probes taken across all three configurations:

| property | result |
|---|---|
| benchmark's copy of the shipped kernel vs. the real one | bitwise equal, 16/16 |
| `fused` vs. baseline | **bitwise equal, 16/16** |
| `packed` (baseline's own negatives replayed) vs. baseline | ≤ 1.12e-05 max abs |
| `packed` repeated twice on identical state | up to 1.48e-05 max abs |
| packed draw hitting an excluded index | 0 / 2,097,152 draws |
| packed draw out of `[0, n_samples)` | 0 / 2,097,152 draws |

The replay error and the repeat-to-repeat difference are the same order of
magnitude, so the packed arm has no systematic bias — the discrepancy is
`index_add_` reassociation, not different math. That is also the problem: the
scatter is non-deterministic on CUDA, so a packed kernel would make
`random_state` stop producing reproducible embeddings.

Quality is unaffected. Neighborhood preservation @15 moves by −1.4% to +1.1%
relative with no consistent sign, which is the noise floor for a different
negative-sampling stream (one embedding per arm per configuration; see
limitations).

## Decision

Measured against the campaign thresholds:

- **Speed (≥10% median wall).** Met at exactly one of three scales: −10.1% at
  n=200k (−10.5% on replication). Missed at n=44.8k (−2.3%) and at n=1.31M
  (−6.8%). The repulsive branch itself is 20% faster at n=44.8k and ~37% faster
  at both larger sizes, but it is only 16–25% of the training step, and the step
  is 81–90% of the wall clock.
- **Memory (≥15% peak).** Missed everywhere: −6.6% / −4.6% / −3.0% of the
  training-step peak. The repulsive phase does shrink 9–19%, but **the attractive
  branch sets the step peak at every scale measured** (5279 vs 3524 MB at 200k;
  51722 vs 33250 MB at 1.3M), so shrinking the negatives cannot move the process
  peak. Most of the small step-peak gain that does appear is not the packed
  kernel at all — it is the `chunk × n_negatives` int64 `neg_indices_` buffer
  (1.57 GB at n=1.3M) that the packed arm never allocates.
- **Quality.** No loss.

Against that, a packed kernel costs: a changed random stream, a
non-deterministic `index_add_` scatter, a dynamic per-iteration shape that forces
a device→host synchronization (`repeat_interleave` with device-resident counts
must read the total), and roughly 45 lines of index machinery in the hot path.

So: REFUTE the packed kernel as a default. This is not a null result — the
n=200k point is a genuine 10%, and a maintainer may still want it — but it does
not clear the bar consistently, and the reproducibility regression is a user-
contract change that a 7% median win does not buy.

Two findings from this work are worth separate consideration and are **not**
proposed here:

1. The repulsive branch builds the `(chunk, n_negatives, n_components)`
   difference tensor twice — once inside `pairwise_distances_indexed` to get the
   distances, once again for the gradient. The `fused` arm removes the second
   build, is **bitwise identical** to the current output, and is 17–21% faster in
   the repulsive branch at every scale. It is not free: holding `diff` and `D`
   simultaneously raises the repulsive-phase peak by 3–5%, and the end-to-end
   effect (−4.4% / −3.4% / +0.5%) is within wall-clock noise at the largest size.
   A decision on that trade-off belongs in its own PR.
2. `PCA` initialization hard-failed once with
   `torch._C._LinAlgError: linalg.svd: The algorithm failed to converge` on the
   already-decorrelated Macosko PCA-50 matrix (job `20736277`), while the same
   input only emitted the cuSOLVER fallback warning on other runs. Flaky, and
   orthogonal to this point.

## Retained benchmark artifact

`benchmarks/umap_negative_sampling.py` — one file, no changes anywhere else.

- Three arms behind `--arms`, written as plain functions of a fitted model so
  they can be compared against one identical state.
- CUDA-event timing for the attractive branch, the repulsive branch and the
  whole step, plus exact per-phase peak allocated bytes.
- Negative-pair density and the widest-per-row-budget statistic, so the
  "truncate to the step maximum" alternative can be re-refuted on other data.
- `--check` runs the gradient probe and the exclusion/range audit of the packed
  draw.
- Downloads the same two single-cell datasets as the existing UMAP benchmark and
  caches them under `--cache-dir`; `--json` writes the full record.

## Validation

- Pre-commit clean on the added file (`ruff`, `ruff-format`, codespell and the
  rest), SLURM job `20736190`.
- Full test suite on 1 × B200: **485 passed, 13 skipped, 6 xfailed, 6 failed**.
  All six failures are the pre-existing `torchdr/tests/test_dataloader.py`
  FAISS CPU-vs-GPU tolerance failures on this hardware (`Greatest absolute
  difference: 0.00390625`), present on `697d7cb` before this branch. This PR adds
  no importable package code, so it cannot affect them.
  - One unrelated flake was seen once and did not reproduce:
    `test_affinity.py::test_normalized_gibbs_affinity[0-sqeuclidean-float32]`
    failed a `tol=1e-5` float32 marginal check in one run, then passed 3/3 in
    isolation and again in the full suite.
- The committed file was re-run end to end after formatting (jobs `20736835`,
  `20736277`); the numbers above come from that exact revision.
- `git diff --check` clean.

## Risks and limitations

- **Single GPU only.** Everything here is world size 1. The repulsive branch is
  chunk-local and per-rank identical, so distributed results should follow the
  same shape — but note the direction: at world size *w* each rank owns *n/w*
  rows, which moves the workload toward the small-*n* regime where packing gives
  nothing, and the packed kernel adds a per-iteration host synchronization that
  would interleave with the gradient all-reduce. Both push the verdict more
  negative, not less, which is why a multi-GPU run was not required to settle it.
- **One seed per configuration.** The quality column is a single embedding per
  arm, not a distribution. It is reported to show the packed arm does not damage
  the embedding, not to resolve sub-percent quality differences.
- At n=1.3M the wall-clock delta (−5.3 s) exceeds the summed step-time delta
  (500 × 7.15 ms = 3.6 s). The residual is outside the timed windows and is
  consistent with allocator pressure from the 1.57 GB `neg_indices_` buffer the
  packed arm never allocates; it was not isolated further.
- The benchmark is 535 lines, larger than the other files in `benchmarks/`,
  because it carries two alternative kernels and a correctness probe alongside
  the measurement. It is standalone and imports nothing new.
- `--dataset zheng` downloads and unpickles a 1.3M-row file from
  `file.biolab.si`, the same source and mechanism as the existing
  `benchmarks/benchmark_umap_single_cell.py`.
- Results are B200-specific. The density and saturation statistics are
  hardware-independent; the timing ratios are not.

### CODE REVIEW

**Comment by huguesva** [comment]
Closing after evidence-based review. The benchmark refutes making packed negative sampling the default: the end-to-end gain is inconsistent across scales, memory savings remain below the acceptance threshold, and reproducibility changes. There is no production change to merge, and retaining a 535-line one-off harness would add maintenance cost. The measurements remain available in this closed PR.
